### PR TITLE
WIP: source: Mysql Source feature

### PR DESCRIPTION
**Pull request by sudharsana-kjl**
This PR hopes to add a feature to work with data from a mysql database.

### CODE REVIEW

File: dffml/source/mysql_source.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    user: str
+    password: str
+    db: str
+    query: str
+
+
+class MysqlSourceContext(BaseSourceContext):
+    async def update(self, repo: Repo):
+        db = self.conn
+        marshall = json.dumps(repo.dict())
```
Comment (pdxjohnny) [inline dffml/source/mysql_source.py]: Ah, we probably want to add another parameter to `MysqlSourceConfig`, which will be the order of things to update, as given in the update query. So for an update query such as:

```sql
INSERT INTO source_data (`key`, `column_a`, `column_b`) VALUES(%s, %s, %s) ON DUPLICATE KEY UPDATE column_a = %s, column_b = %s;
```

We will need a parameter which is the feature names in order.

```console
-model-features feature_a feature_b
```

Which maps to the columns they are stored in, `column_a` and `column_b`.  You'll probably need an `collections.OrderedDict`.

File: CHANGELOG.md [hunk 2]
```diff
@@ -16,6 +16,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   DFFML models and operations.
 - Classification type parameter to DNNClassifierModelConfig to specifiy data
   type of given classification options.
+- Mysql Source feature to work with data from a mysql database
```
Comment (pdxjohnny) [inline CHANGELOG.md]: `Mysql Source feature` should be `Mysql Source`

**Comment by pdxjohnny** [comment]
```python
import unittest

from dffml.util.testing.source import SourceTest
from dffml.util.asynctestcase import AsyncTestCase

from dffml.source.mysql import MysqlSourceConfig, MysqlSource


class TestMysqlSource(SourceTest, AsyncTestCase):
    async def setUpSource(self):
        return MysqlSource(
            MysqlSourceConfig(... parameters, get from os.getenv ...)
            # user=os.getenv("MYSQL_USER")
            # password=os.getenv("MYSQL_PASSWORD")
        )

    async def test_update():
        await super().test_update()
```

We will start a mysql database within travis CI and set environment variables the same way we start the database, aka, same `MYSQL_USER`, `MYSQL_PASSWORD` from [docker hub docs](https://hub.docker.com/_/mysql): @pdxjohnny will do this for travis.

File: dffml/source/mysql_source.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    async def __aenter__(self) -> "MysqlSource":
+        self.pool = await aiomysql.create_pool(
+            host=self.config.host,
+            port=self.config.port,
+            user=self.config.user,
+            password=self.config.password,
+            db=self.config.db,
+        )
+        self.__db = self.pool.acquire()
+        self.db = await self.__db.__aenter__()
```
Comment (pdxjohnny) [inline dffml/source/mysql_source.py]: This is where the connection is being established.

**Comment by pdxjohnny** [comment]
To have `aiomysql` installed for testing, add to setup.py

```python
    # ...
    packages=find_packages(),
    install_requires=[
        "aiomysql>=Go find the latest version number of PyPi and put it here",
    ],
    include_package_data=True,
    # ...
```

File: dffml/source/mysql_source.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        repo = Repo(src_url)
+        db = self.conn
+        await db.execute(query, (src_url,))
+        dump = await db.fetchone()
+        if dump is not None and dump[0] is not None:
+            repo.merge(Repo(src_url, data=json.loads(dump[0])))
+        return repo
+
+    async def __aenter__(self) -> "MysqlSourceContext":
+        self.__conn = self.parent.db.cursor(aiomysql.DictCursor)
```
[the remark this answers is not in the document]
Comment (sudharsana-kjl) [reply dffml/source/mysql_source.py]: I've modified it now. :+1:

File: dffml/source/mysql_source.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+                        values += repo.data.__dict__[key] + ","
+                        column_value_pair[column_name] = repo.data.__dict__[
+                            key
+                        ]
+                        update_query += column_name + ","
+            update_query = update_query[:-1] + ")"
+            values = values[:-1] + ")"
+            update_query += " values " + values + "on duplicate key update "
+            for key, value in column_value_pair.items():
+                update_query += key + "=" + str(value) + ","
```
Comment (pdxjohnny) [inline dffml/source/mysql_source.py]: Building SQL queries can be dangerous, see [SQL Injection](https://www.w3schools.com/sql/sql_injection.asp) for more info. The big thing to learn about SQL injection is that you need to bind your parameters. 

In the following example `feature_cols` is your `model_columns`.

https://github.com/intel/dffml/blob/86a0168fdd42e59fee427bb6569ed81cd12bb9e0/examples/source/custom_sqlite.py#L19-L28

However sqlite uses `?` to bind parameters, where as `aiomysql` uses `%s`

https://github.com/intel/dffml/blob/86a0168fdd42e59fee427bb6569ed81cd12bb9e0/examples/maintained/demoapp/source.py#L29-L33

The last argument to the `execute` call is a `tuple` of parameters which are associated with each `%s`, similar to string formatting.

File: tests/source/test_mysql.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+            MysqlSourceConfig(
+                host="localhost",
+                port=3306,
+                user="root",
+                password="ABcd1234",
+                db="dffml_source",
+                table="repo_data",
+                repo_query="",
+                update_query="",
+                repos_query="",
```
Comment (pdxjohnny) [inline tests/source/test_mysql.py]: This won't work unless you set `repo_query, update_query, repos_query`

File: dffml/source/mysql_source.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+                            "features": {
+                                key.replace("feature_", ""): value
+                                for key, value in row.items()
+                                if key.startswith("feature_")
+                            },
+                            "prediction": {
+                                key.replace("prediction_", ""): value
+                                for key, value in row.items()
+                                if key.startswith("prediction_")
+                            },
```
[the remark this answers is not in the document]
Comment (sudharsana-kjl) [reply dffml/source/mysql_source.py]: Ohh in  SourceTest as well as in CSVSource, prediction takes two values:
https://github.com/intel/dffml/blob/86a0168fdd42e59fee427bb6569ed81cd12bb9e0/dffml/source/csv.py#L87-L95

Should i make the changes here as well?

File: dffml/source/mysql_source.py [hunk 8]
```diff
@@ -16,100 +16,62 @@ class MysqlSourceConfig(BaseConfig, NamedTuple):
...
+                Repo(
+                    row["src_url"],
+                    data={
+                        "features": {
+                            key.replace("feature_", ""): value
+                            for key, value in row.items()
+                            if key.startswith("feature_")
+                        },
+                        "prediction": {
+                            key.replace("prediction_", ""): value
```
Comment (pdxjohnny) [inline dffml/source/mysql_source.py]: This looks good.

**Comment by sudharsana-kjl** [comment]
For testing purpose, SSL certificate was generated using `openssl`. 
```
openssl genrsa 2048 > private.pem
openssl req -x509 -days 1000 -new -key private.pem -out ca.pem
```
While generating cert, it'll ask for details such as country code, state code, etc. I've generated the cert used in tests/ssl_resources by giving only country code as 'IN', rest of the fields are left blank.

File: dffml/source/mysql_source.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (pdxjohnny) [reply dffml/source/mysql_source.py]: You're right, I was wrong about this

**Comment by pdxjohnny** [comment]
I'm working on fixing the merge test. I'll fix that for JSONSource, modify things here, and merge it all soon.

**Comment by pdxjohnny** [comment]
I'll get to adding this to travis soon

**Pull request merged**